src/main/java/soot/dexpler/DexFillArrayDataTransformer.java

```diff
@@ -0,0 +1,138 @@
+ * Instead {@link UntypedConstant} constants are used. These constants are processed by this transformer and get their final
+ * type.
+ * 
+ *
+ * @author Jan Peter Stotz
+ *
+ */
+public class DexFillArrayDataTransformer extends BodyTransformer {
+  private static final Logger logger = LoggerFactory.getLogger(DexFillArrayDataTransformer.class);
+
+  public static DexFillArrayDataTransformer v() {
+    return new DexFillArrayDataTransformer();
+  }
+
+  protected void internalTransform(final Body body, String phaseName, Map<String, String> options) {
+    final ExceptionalUnitGraph g = ExceptionalUnitGraphFactory.createExceptionalUnitGraph(body, DalvikThrowAnalysis.v());
+    final LocalDefs defs = G.v().soot_toolkits_scalar_LocalDefsFactory().newLocalDefs(g);
+
+    for (Iterator<Unit> unitIt = body.getUnits().snapshotIterator(); unitIt.hasNext();) {
+      Unit u = unitIt.next();
+      if (!(u instanceof AssignStmt)) {
+        continue;
+      }
+      AssignStmt ass = (AssignStmt) u;
+      Value rightOp = ass.getRightOp();
+      if (rightOp instanceof UntypedConstant) {
+        Value left = ass.getLeftOp();
+        if (left instanceof ArrayRef) {
+          ArrayRef leftArray = (ArrayRef) left;
+
+          Local l = (Local) leftArray.getBase();
+          List<Unit> assDefs = defs.getDefsOfAt(l, ass);
+          List<Type> arrayTypes = new LinkedList<>();
+          for (Unit d : assDefs) {
+            if (d instanceof AssignStmt) {
+              AssignStmt arrayAssign = (AssignStmt) d;
+              Value source = arrayAssign.getRightOp();
+              if (source instanceof NewArrayExpr) {
+                NewArrayExpr newArray = (NewArrayExpr) source;
+                arrayTypes.add(newArray.getBaseType());
+                continue;
+              }
+              if (source instanceof InvokeExpr) {
+                InvokeExpr invExpr = (InvokeExpr) source;
+                Type aType = invExpr.getMethodRef().getReturnType();
+                if (!(aType instanceof ArrayType)) {
+                  throw new InternalError("Failed to identify the array type. The identified method invocation "
+                      + "does not return an array type. Invocation: " + invExpr.getMethodRef());
+                }
+                arrayTypes.add(((ArrayType) aType).getArrayElementType());
+                continue;
+              }
+              throw new InternalError("Unsupported array definition statement: " + d);
+            }
+          }
+          if (arrayTypes.isEmpty()) {
+            throw new InternalError("Failed to determine the array type ");
+          }
+          if (arrayTypes.size() > 1) {
+            arrayTypes = arrayTypes.stream().distinct().collect(Collectors.toList());
+            if (arrayTypes.size() > 1) {
+              logger.warn("Found multiple possible array types, using first ignoreing the others: {}", arrayTypes);
```

Comment:
Wouldn't it make sense to find a least common ancestor (lcas) in hthe type hierarchy? Soot already uses this strategy in the type assigner and there is already code for it (look for ``lcas``).

Reply:
@StevenArzt Are you sure this is really necessary? As you noticed in #1806 fill-array-data only works for primitive types like byte, char, short, int, long, float, double. None of these types have a least common ancestor. Therefore if  `arrayTypes` would contain such a type no matter what we do we will end up in an exception.

Reply:
When I was asking about objects I wasn't very familiar with the `fill-array-data` command. Theoretically it would be possible to use it to fill an object array, but the array can only held references to an object and not the object itself.
Thus it would be necessary to save the object references ("pointers") in the array data block. As this block is defined at compile-time you would have to know the object reference "pointers" at compile-time.

Also a quick practical test uncovered that the Android verifier rejects if you try to apply `fill-array-data` onto an `Object[]` array:

> Error from logcat: invalid fill-array-data with component type

Reply:
OK, great. That makes the Dexpler implementation a lot easier.